userguide/tutorials/entitlement_plugin.adoc

```diff
@@ -0,0 +1,40 @@
+= Developing an Entitlement Plugin
+
+[[overview]]
+== Overview
+
+=== EntitlementPluginApi
+
+The https://github.com/killbill/killbill-plugin-api/blob/master/entitlement/src/main/java/org/killbill/billing/entitlement/plugin/api/EntitlementPluginApi.java[EntitlementPluginApi] exists to develop `entitlement plugins`. Those plugins are called from the core entitlement system to provide a way for the plugins to intercept any of the subscription operations (e.g intercept creation/cancelation, ... for subscriptions). This is similar to the https://docs.killbill.io/latest/payment_control_plugin.html[payment control plugin api], but instead of intercepting payment calls, this intercepts subscriptions related calls.
+
+
+Each of the APIs take the following parameters:
+
+* `context`: Provides the full `EntitlementContext` associated with the (set of) subscription(s) (account, operation type, ...)
+* `properties`: A list of `PluginProperty` that are passed all the way from the client making the API call to the plugin. Each plugin can therefore be controlled through properties to achieve the desired result and provide additional state (in addition to the basic subscription state) to the plugin that way.
+
+Implementers of such plugins need to implement the following apis:
+
+* `priorCall`: This is the method that will be invoked prior doing any subscription operation. This method will return a `EntitlementContext` that will allow to:
+** abort the call if desired 
+** adjust the parameters for the call (e.g use a different `Plan` to create a subscription, or creates a price override, ...)
+* `onSuccessCall`: This is invoked after the subscription operation was executed successfully. This method will return a `OnSuccessEntitlementResult` which is currently set to an empty result
+* `onFailureCall`: This is invoked after the subscription operation failed to execute successfully. This method will return a `OnFailureEntitlementResult` to which is also set to an empty result
+
+Multiple of such plugins can be registered in the system. Each plugin will be invoked in order and if a plugin decides to modify the list of `PluginProperty`, subsequent plugins then see that modifed list of properties. Obviously, this is a very powerful mechanism which should not be abused (to avoid having to debug a *too* complicated system).
```

Comment:
I would remove this paragraph in light of the next section.

Reply:
Fixed in https://github.com/killbill/killbill-docs/pull/281/commits/d34bd5388ad7641ec634de270cd38843ac77f21c